### PR TITLE
fix(phpstan): fix phpstan issues

### PR DESCRIPTION
**Pull request by blackshadev**
## What 

Related to #379

I resolved some issues related to phpstan. I moved some ignore patterns around, removed unused ignore patterns, fixed some issues.

Here is the full breakdown:

Some are ignored in the "normal" phpstan config file like others. 
- forbidden `exec` in `src/Tempest/Console/src/Terminal/Terminal.php` . Which seems as expected and other exec calls are also ignored this way
- `src/Tempest/Http/src/Exceptions/exception.php` seems to have been moved, so I changed the path for the ignore rule

Others where ignore patterns which were no longer required:
- `NullShellExecutor.php`
- `CircularWithInitializer[ABC].php`
- `ModelDefinition.php`
- I excluded the whole `tests/Integration/View/blade/cache/**.php` path instead of ignoring errors from it.

Lastly I fixed some actual  issues:
- issues in `Status.php`:
  -  Some cases were missing in description, this would cause issues, so jeeej for phpstan spotting it.
  - In some condition a "sanity" check was done on the status value, like `status >= 100` and  `status < 600` . But phpstan doesn't like sanity, it likes strict typing ;) 
- `Route.php`: Fixed some logic you cannot apply a template in the middle of any class. In this case I didn't think it was needed anyway.
- `HttpExceptionHandler.php` :  added a "hack" to ensure `getCodeSample` get's referenced. If you don't like it I can always re-add it to the ignore list
- `HttpExceptionHandler.php`: str_pad got used incorrectly, this actually seemed like a bug? so jeeej for phpstan spotting it.

## Pending issues
- `HttpExceptionHandler.php` :  added a "hack" to ensure `getCodeSample` get's referenced. If you don't like it I can always re-add it to the ignore list
- The whole `A should be final` thing seems weird. I get that you would want to enforce final, but something that is explicitly marked as abstract should be the exception right. From my knowledge phpstan should not report this on abstract classes.

### CODE REVIEW

File: src/Tempest/Http/src/Status.php
Comment (aidan-casey) [inline src/Tempest/Http/src/Status.php:1]: I like having the additions of the descriptions, but I think we should put this in a separate PR.

File: src/Tempest/Http/src/Status.php [hunk 1]
```diff
@@ -157,7 +159,7 @@ public function description(): string
 
     public function isInformational(): bool
     {
-        return $this->value >= 100 && $this->value < 200;
+        return $this->value < 200;
```
Comment (aidan-casey) [inline src/Tempest/Http/src/Status.php:162]: What's the advantage of this change?

File: src/Tempest/HttpClient/src/Driver/Psr18Driver.php [hunk 2]
```diff
@@ -28,17 +28,15 @@ public function __construct(
...
+        $psrRequest = $this->convertTempestRequestToPsrRequest($request);
+
         return $this->convertPsrResponseToTempestResponse(
-            $this->sendRequest($request)
+            $this->sendRequest($psrRequest)
         );
     }
 
-    public function sendRequest(Request|RequestInterface $request): ResponseInterface
+    public function sendRequest(RequestInterface $request): ResponseInterface
```
Comment (aidan-casey) [inline src/Tempest/HttpClient/src/Driver/Psr18Driver.php:38]: `Request|RequestInterface` is necessary to be PSR compliant.

File: tests/Fixtures/Views/rawAndEscaping.php [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 use Tempest\View\GenericView;
 
-/** @var GenericView $this */ ?>
+/** @var GenericView|stdClass $this */ ?>
```
Comment (aidan-casey) [inline tests/Fixtures/Views/rawAndEscaping.php:5]: @brendt - Can you just confirm this one?

**Review by aidan-casey — CHANGES_REQUESTED** [review]
Thanks, @blackshadev! Just a few comments.

File: src/Tempest/Http/src/Status.php
Comment (blackshadev) [reply src/Tempest/Http/src/Status.php:1]: These added descriptions are required for phpstan, phpstan actually complained we were missing some cases ;) so why a separate PR?

File: src/Tempest/Http/src/Status.php (hunk 1 above)
Comment (blackshadev) [reply src/Tempest/Http/src/Status.php:162]: This is the part about phpstan doesn't like sanity checks
> In some condition a "sanity" check was done on the status value, like status >= 100 and status < 600 . But phpstan doesn't like sanity, it likes strict typing ;)

The reason why phpstan errors on this is because it knows $value can be a certain set of values, which start above 100. Meaning `$this->value >= 100`  will always be true. And that is the phpstan error we got.

File: src/Tempest/HttpClient/src/Driver/Psr18Driver.php (hunk 2 above)
Comment (blackshadev) [reply src/Tempest/HttpClient/src/Driver/Psr18Driver.php:38]: I am very certain that is not true. `Psr\Http\Client\ClientInterface`  has an send method with this signature 
`public function sendRequest(RequestInterface $request): ResponseInterface;`. The `Request` is a `Tempest\Http\Request` which doesn't seem like it has anything to do with PSR.

The phpstan error actually said this, the method was required to be `sendRequest(RequestInterface $request): ResponseInterface` and this actually shouldn't accept `Tempest\Http\Request` which we did provide and this caused an error.

File: tests/Fixtures/Views/rawAndEscaping.php (hunk 3 above)
Comment (blackshadev) [reply tests/Fixtures/Views/rawAndEscaping.php:5]: for clarification it solves the error caused by `<?= $this->property ?>` . `property` is a property of `GenericView`  thus this will error. By unioning `stdClass` we can add / access any property without phpstan complaining.

Comment (brendt) [reply tests/Fixtures/Views/rawAndEscaping.php:5]: Wouldn't PHPStan know that there's a magic getter on GenericView?

Comment (blackshadev) [reply tests/Fixtures/Views/rawAndEscaping.php:5]: It knows, but you will still get the `Access to an undefined property Tempest\View\GenericView::$property.` error because you should than still define which properties one can get (instead of just any). "normally" you whouldn't be able to get just about anything, there would still be a limit. But we do not know this limit in this case.

Comment (brendt) [reply tests/Fixtures/Views/rawAndEscaping.php:5]: Gotcha, we were ignoring the error before — right? I'm fine with this docblock, seems like a good solution :)

File: src/Tempest/HttpClient/src/Driver/Psr18Driver.php (hunk 2 above)
Comment (brendt) [reply src/Tempest/HttpClient/src/Driver/Psr18Driver.php:38]: It could actually be that we allowed `Request|RequestInterface` before, then refactored some stuff, and now only `RequestInterface` is needed. Part of the alpha phase 😅 If all tests are green and PHPStan is as well, then I'm fairly certain we're ok to drop `Request`

File: src/Tempest/Http/src/Status.php (hunk 1 above)
Comment (brendt) [reply src/Tempest/Http/src/Status.php:162]: But how would this be correct? `Status::CONTINUE->isInformational()` will now return `true` while it should return `false`?

File: src/Tempest/HttpClient/src/Driver/Psr18Driver.php (hunk 2 above)
Comment (aidan-casey) [reply src/Tempest/HttpClient/src/Driver/Psr18Driver.php:38]: Naw, we have to keep this in unless we are going to change some design decisions.

Originally I had it as a PSR only client, but @brendt wanted us to be able to send Tempest request objects as well. This change was made per his request [here](https://github.com/tempestphp/tempest-framework/pull/157#pullrequestreview-1912845989).

**Comment by brendt** [comment]
FYI 

> tests/Fixtures/Views/rawAndEscaping.php

That one can be removed :)

File: src/Tempest/Http/src/Status.php (hunk 1 above)
Comment (blackshadev) [reply src/Tempest/Http/src/Status.php:162]: I want to believe you , but... `Status::CONTINUE = 100` correct? meaning the old code `$this->value >= 100 && $this->value < 200`  wil evaluate to `true` . The new code of `$this->value < 200` also returns `true`. If it shouldn't return `true` I can change the expression. But functionally my code is the same as the old one. 

PHPStan noted that because Status starts at `100` the expression `$this->value >= 100` is always true. That is why we can remove it. If we should is another topic. But in my eyes, we would never have a status sub 100, so why not.

**Comment by blackshadev** [comment]
> > tests/Fixtures/Views/rawAndEscaping.php

Rebased and removed, thx!

**Comment by blackshadev** [comment]
I will fix the pipeline later today. It seems that some merged fixed (or moved) some phpstan issues which causes it to fail

File: src/Tempest/Http/src/Status.php (hunk 1 above)
Comment (brendt) [reply src/Tempest/Http/src/Status.php:162]: Ooooh I'm a total fool… I read it as `>=200 and <300` no idea why 😅

Yeah, totally fine this change. Sorry for the noise

**Comment by brendt** [comment]
Thanks!
